File: dart/design_patterns/lib/behavioral/iterator.dart
// Intent: Provide a way to access elements of a collection
// sequentially without exposing its underlying representation.

// Iterator
// Defines the interface for accessing and traversing elements.

abstract interface class MyIterator {
  bool get hasNext;
  int? next();
}

// Aggregate
// Defines the interface for creating an iterator.

abstract interface class MyAggregate {
  MyIterator createIterator();
}

// Concrete Iterator
// Implements the iterator interface and tracks the current position.

class MyCollectionIterator implements MyIterator {
  final List<int> _items;
  int _currentIndex = 0;

  MyCollectionIterator(this._items);

  @override
  bool get hasNext => _currentIndex < _items.length;

  @override
  int? next() {
    if (!hasNext) return null;
    final item = _items[_currentIndex];
    _currentIndex++;
    return item;
  }
}

// Concrete Aggregate
// Implements the aggregate interface and returns a concrete iterator.

class MyCollection implements MyAggregate {
  final List<int> _items;

  MyCollection(this._items);

  @override
  MyIterator createIterator() => MyCollectionIterator(_items);
}

// Usage
// Iterate through the collection without knowing its internal structure.

void iteratorExample() {
  final items = [1, 2, 3, 4, 5];
  final myCollection = MyCollection(items);
  final iterator = myCollection.createIterator();

  while (iterator.hasNext) {
    print(iterator.next());
  }
}

void main() {
  iteratorExample();
}

File: swift/DesignPatterns/Behavioral/Iterator.swift
//
//  Iterator.swift
//  DesignPatterns
//
//  Created by Soe Min Thein on 24/03/2026.
//

import Foundation

protocol MyIterator {
    var hasNext: Bool { get }
    mutating func next() -> Int?
}

protocol MyAggregate {
    func makeIterator() -> MyIterator
}

struct MyCollectionIterator: MyIterator {
    private let items: [Int]
    private var currentIndex = 0
    
    init(items: [Int]) {
        self.items = items
    }
    
    var hasNext: Bool {
        currentIndex < items.count
    }
    
    mutating func next() -> Int? {
        guard hasNext else { return nil }
        let item = items[currentIndex]
        currentIndex += 1
        return item
    }
}

struct MyCollection: MyAggregate {
    private let items: [Int]
    
    init(items: [Int]) {
        self.items = items
    }
    
    func makeIterator() -> MyIterator {
        MyCollectionIterator(items: items)
    }
}

func iteratorExample() {
    let items = [1, 2, 3, 4, 5]
    let myCollection = MyCollection(items: items)
    var iterator = myCollection.makeIterator()
    
    while iterator.hasNext {
        print(iterator.next() as Any)
    }
}
